Extract db-related methods from baseTrie

s1na opened the pull request:
This PR extracts db-related methods such as `getRaw`, `putRaw` from `BaseTrie` and moves them to a new class `DB`. It also modifies how checkpointing is done in `CheckpointTrie`.

The goal was to do these two modifications according to [this comment](https://github.com/ethereumjs/merkle-patricia-tree/issues/67#issuecomment-454329414), but after implementation I figured out some problem with that design. I experimented with a few other designs, but they all seemed hacky like the current checkpointing method. This PR is one of those methods which I liked a bit more than others. But if you think it doesn't improve on the status quo, we can close this PR.

To achieve checkpointing we somehow have to make the original trie read-only, and perform all subsequent writes to a temporary (scratch) db. On `revert` the scratch db is dropped and trie root is set to the checkpointed root. On `commit` nodes in the scratch db which correspond to the current trie (with current root) have to be written to the main persistent db.

For this purpose, this PR introduces a `ScratchDB`, which is an in-memory `DB` with a backend (pointer to main persistent db). On checkpoint, `CheckpointTrie` replaces the db that `BaseTrie` uses with an instance of `ScratchDB`, therefore all writes are written to the scratch. If a key is not found in the `ScratchDB`, it tries to fetch it from its `backend`.

I also have some questions regarding `getRaw` and `putRaw`. My impression from reading the existing code is that `putRaw` will write to the persistent DB even if trie is during a checkpoint. Is that intentional? If so, should `getRaw` and `delRaw` also be done on persistent db regardless of checkpointing? We should probably highlight this in documentation, or rename them to make this side-effect clear.

holgerd77 commented:
Whew, at least tests are already passing, great! 👍 😄

holgerd77 commented:
Think I like this approach, this get's definitely more readable and distinctly sorted out regarding the functionality respectively responsibility.

I wonder if we should generally switch to a more directory-structured file structure (and also take on some breaking import changes as a consequence), suggestion:

- `index.js`
- `trie` directory
  - `base.js``
  - `checkpoint.js`
  - `secure.js`
  - `node.js`
  - `proof.js`
- `db` directory
  - `db.js`
  - `scratch.js`

Not completely sure what to do with the other files `taskExecutor` and `readStream`. Does this generally make sense? Or we keep the trie classes in the main directory and just introduce one for `db`?

holgerd77 commented:
Find this file and naming mix generally still too messy...

holgerd77 commented:
Updated the directory structure above with latest reflections/ideas on this.

Side node: we can also address this in a separate PR.

holgerd77 commented:
Tested the [basic usage](https://github.com/ethereumjs/merkle-patricia-tree#usage) example, worked.

holgerd77 commented on src/checkpointTrie.js:
```diff
@@ -71,7 +76,8 @@ module.exports = class CheckpointTrie extends BaseTrie {
    * Reverts the trie to the state it was at when `checkpoint` was first called.
-   * If during a nested checkpoint, only sets parent as current checkpoint.
+   * If during a nested checkpoint, sets root to most recent checkpoint, and sets
+   * parent checkpoint as current.
```
This is just a better explanation of existing functionality, good.

holgerd77 commented on src/checkpointTrie.js:
```diff
@@ -111,60 +118,50 @@ module.exports = class CheckpointTrie extends BaseTrie {
+    if (this.isCheckpoint) return this._overridePutRaw(key, val, cb)
+    this.db.put(key, val, cb)
+  }
+
```
To put this together and to your question, just summing up to make sure I get everything right: Before `putRaw` was ignoring checkpoint mechanism and would directly write to the permanent `DB`. This PR changes this behavior and `putRaw` writes to the scratch DB. If one wants to get "the old" functionality, one would directly use the method from `DB` to perform this.

Is this correct?

If so, I very much like this change, since this makes it very much more transparent and should avoid false usage of the API.

holgerd77 replied:
I think when this is moved to `DB` and we decide to remove the methods from the trie class it becomes much clearer that the functionality is actually working on the `raw` un-checkpointed DB (and, to make this even clearer, we can also state additionally in the JSDoc comments).

So I think we should be good with this for now, since we are not taking away functionality. If people have some need for raw methods on the scratch DB, we can always think this properly through and add this later.

holgerd77 commented on src/baseTrie.js:
```diff
@@ -50,7 +53,8 @@ module.exports = class Trie {
 
-    this.putRaw = this._putRaw
+    this.getRaw = this.db.get.bind(this.db)
+    this.delRaw = this.db.del.bind(this.db)
```
Update: after some more digging into the code and having made this `putRaw` - scratch vs permanent DB write - comment: this changes a bit the notion of this, updated question: is having these `raw` methods being used through `trie` and transparently working rather on scratch or permanent DB useful? Or should we rather go one step further and remove this here and eventually completely disallow the old raw `level` passing constructor here and tell people to switch to instantiate with `DB`?

s1na replied:
I'm not sure to be honest, this is my question too. Do you know how or if these `raw` methods are being used in the wild? Maybe that'd help us in making the decision. I'm more inclined towards removing the methods (but maybe because I'm not aware of its use-cases)

holgerd77 replied:
Since the use cases respectively the functionality provided (writing to the permanent DB) stay the same when invoked through `DB` - I think we can remove these here and tell people to use through the DB class.

This would also play well with the suggestion above to force the instantiation with `DB`.

holgerd77 commented on src/checkpointTrie.js:
```diff
@@ -111,60 +118,50 @@ module.exports = class CheckpointTrie extends BaseTrie {
+    trie.db = scratch
     trie._scratch = scratch
     return new ScratchReadStream(trie)
   }
```
Just wondering: is there any public use case for this method so  this is not marked as `private` in some way? Since it is taking `scratch` as a parameter?

s1na replied:
I kept it public only because it was public before. This could potentially be used to get modifications to the trie since checkpoint as a read stream. Not sure if there's any reason why this'd be helpful, or if anyone is using this.

s1na replied:
Did a quick search, doesn't seem to be used. Made it private.

holgerd77 replied:
Ok, fine as well.

holgerd77 commented on src/checkpointTrie.js:
```diff
@@ -111,60 +118,50 @@ module.exports = class CheckpointTrie extends BaseTrie {
-    this._putRaw = this.putRaw
-    this.putRaw = this._overridePutRaw
+    this._scratch = new ScratchDB(this._mainDB)
+    this.db = this._scratch
```
We have some naming confusion between `DB` the class and `DB` the level instance, respectively this would be more easy to grasp if we would differentiate here.

Have no brilliant idea for a naming scheme here (prefixing all `level` stuff with `level` might also not be the best idea since eventually we want to support further types of storage in the future).

One idea: prefix with `Trie` for the class versions, since these have some trie related wrapping changes, so e.g. `TrieDB.js` and `this.trieDB`. Hmm, but doesn't like this yet too much myself though.

Any thoughts?

Follow-up thought: maybe this would be good to be named such as `DB` the class remains as simple as currently be since this is the main exposed part of the API.

holgerd77 replied:
Yeah, that would make sense.

holgerd77 commented on src/checkpointTrie.js:
```diff
@@ -111,60 +118,50 @@ module.exports = class CheckpointTrie extends BaseTrie {
-      async.map(this._putDBs, flushScratch, cb)
     } else {
-      cb()
+      async.nextTick(cb)
```
Is this `async.nextTick()` actually fixing this proxy method thing?

holgerd77 replied:
Generally went through the whole new checkpoint process and checked if this works and is consistent, looks good.

s1na replied:
The `nextTick` is there mainly to prevent ["releasing Zalgo"](https://blog.izs.me/2013/08/designing-apis-for-asynchrony) (see https://github.com/ethereumjs/ethereumjs-vm/issues/421).

holgerd77 replied:
Ah, thanks.

holgerd77 commented on src/scratch.js:
```diff
@@ -0,0 +1,36 @@
+
+    asyncFirstSeries(getDBs, dbGet, cb)
+  }
+}
```
Had to do some longer look to understand how this is working but getting it now, had some trouble to grasp which DB is used on which occasion! 😄

Looks good and really elegant way to achieve. Does this have any performance implications? Did you measure?

s1na replied:
Yeah it's still not as clear... Maybe we can iterate on this further.

No I didn't, but intuitively this shouldn't affect performance by much, as the approach is really similar to what was there before, and the changes are mostly structural.

s1na commented:
Thanks for the extensive review! Will go through your comments now.

Re. the file structure, agree and what you proposed looks good for the current state. Only maybe we can wait a bit more before solidifying the structure, as there's another PR incoming, and then there's transition to typescript which could offer new opportunities for structuring.

s1na commented on src/baseTrie.js:
```diff
@@ -26,11 +29,11 @@ module.exports = class Trie {
+      this.db = db
+    } else {
+      this.db = new DB(db)
+    }
```
You're right, we should probably keep `DB` internal and have the public interface work with `level` instance.

holgerd77 replied:
I have actually a tendency to the other way around, to enforce instantiation by using the `DB` wrapper class. Then we already have some good preparation if we want to add new storage options beside from `level` DB in the future. What do you think? I am in the breaking-changes mood! 😄

s1na commented on src/baseTrie.js:
```diff
@@ -178,60 +156,11 @@ module.exports = class Trie {
-
-    async.each(this._putDBs, del, cb)
-  }
-
```
On checkpoint, `CheckpointTrie` modified `this._putDBs` to control where to write the temporary modifications. These proxy methods go over every `_putDBs` and write the key value.

holgerd77 commented:
Makes very much sense to let the file structure changes sink in a bit more and do this along `TypeScript` transition. Have opened a separate issue https://github.com/ethereumjs/merkle-patricia-tree/issues/75 where we can continue discussion respectively track the latest state of proposals on that.

s1na commented:
Tried to apply your suggestions. Hope I didn't miss anything. `Raw` methods are dropped (and their test file). Also made `createScratchReadStream` private. `DB` uses `this._leveldb` internally to refer to the leveldb instance.

holgerd77 commented:
Hmm, can't we use the tests on the `DB` class?

There is still a linting error preventing the CI run to pass.

holgerd77 commented on src/baseTrie.js:
```diff
@@ -29,11 +29,7 @@ module.exports = class Trie {
-    } else {
-      this.db = new DB(db)
-    }
+    this.db = db || new DB()
```
Can you update the API docs here as well?

holgerd77 commented:
Ready for re-review? 😄

s1na commented:
Yep :smile: I was just waiting to see if CI passes.

I've replaced the `rawOps` tests with unit tests for `db.js`, `scratch.js` and added some additional cases to checkpointing tests (moved them to a separate file `checkpoint.js`).

holgerd77 commented:
Just as some note, no need to change in this round: maybe we should name `DB` even more generic to `Storage` or something, then this is already better named/prepared for some future storage backend exchange? The exposed `DB` API is pretty limited in scope (`get`, `put`, `batch`, `copy`), so it should be not too far-fetched to easily allow switching here in the future with other-backend-targeting wrapper classes?

holgerd77 commented:
Also just as some note and not a PR blocker: these re-generated docs are better than nothing, but after all these changes we have to significantly restructure the docs to reflect the directory/file structure + API exposed.

holgerd77 commented:
Maybe if we just release after `TypeScript` transition, that we also be the point for the docs update.

holgerd77 commented:
Also some note: we had some extensive discussion with Alex when he was doing the changes around changing or not changing the API by the decision where to put the distribution files (so that `require('secure.js')` would not have to be changed to `require('dist/secure.js')`. We ended up to put these all in the root directory and exclude in `.gitignore`.

However after some time dealing with this I have to say this is a total mess. On `TypeScript` transition we should also switch to just the standard way we are doing on other libraries and put everything in the `dist` folder. There will be several breaking changes anyhow so this will be a good occasion to update.

s1na commented:
Weird that coverage changed at the end! but luckily it helped us find a bug! A few points:

1. Added a simple put/get test for `secure`, as well as a `copy` test.
2. Added `copy` tests for `checkpointTrie`, both before any checkpointing and after. The after case didn't work. I've now fixed the problem.
3. The `copy` method of all tries doesn't actually copy the `leveldb`. This means if they copy trie and use both instances in parallel, they're using the same `leveldb`. Tried to add this to docs. I don't know however if this is desired (the alternative of copying whole `leveldb` is also not elegant).
4. I had forgotten to mention that I've changed the `DB` api. Its methods only accept `Buffer` type now and throw an error if input is not a `Buffer`. I thought this is a good opportunity to do this, as we're breaking API anyway :)

holgerd77 commented:
On 3.: Think you have some better oversight then me on ways to organize this, feel free to realize the one you think is most appropriate. Just this throwing everything into root dir makes no sense.
